cql3/user_types.cc

```diff
@@ -295,4 +295,25 @@ void user_types::setter::execute(mutation& m, const clustering_key_prefix& row_k
     }
 }
 
+void user_types::setter_by_field::execute(mutation& m, const clustering_key_prefix& row_key, const update_parameters& params) {
+    assert(column.type->is_user_type() && column.type->is_multi_cell());
+
+    auto value = _t->bind(params._options);
+    if (value == constants::UNSET_VALUE) {
```

Comment:
Why not bind_and_get? It's faster.

I'm missing how we'll get UNSET_VALUE here, if _t is a marker.

Reply:
> I'm missing how we'll get UNSET_VALUE here, if _t is a marker.

Indeed, I didn't catch that. In `constants::marker::bind`:

```        virtual ::shared_ptr<terminal> bind(const query_options& options) override {
            auto bytes = bind_and_get(options);
            if (!bytes) {
                return ::shared_ptr<terminal>{};
            }
            return ::make_shared<constants::value>(std::move(cql3::raw_value::make_value(to_bytes(*bytes))));
        }
```

Instead of checking `!bytes`, shouldn't we check for `unset_value` vs `null_value` and return constants::UNSET_VALUE in the unset vase?

Because right now I see that e.g. maps have the same problem:
```
cqlsh:a> create table tm (a int primary key, b map<int, int>);
cqlsh:a> update tm set b[0] = 0 where a = 0;
cqlsh:a> select * from tm;

 a | b
---+--------
 0 | {0: 0}

```
then from python driver:
```
    session.execute(session.prepare('update tm set b[0] = ? where a = 0'))
```

and we're left with:
```
cqlsh:a> select * from tm;

 a | b
---+---
```

and this happens because:
```
void
maps::setter::execute(mutation& m, const clustering_key_prefix& row_key, const update_parameters& params, const column_definition& column, ::shared_ptr<terminal> value) {
    if (value == constants::UNSET_VALUE) {
        return;
    }
```

In Cassandra it works as expected (the python driver part doesn't have any effect).

Reply:
Actually, my mistake, I was looking at wrong code in maps. The problem is in `setter_by_key` of course:
```
void
maps::setter_by_key::execute(mutation& m, const clustering_key_prefix& prefix, const update_parameters& params) {
    using exceptions::invalid_request_exception;
    assert(column.type->is_multi_cell()); // "Attempted to set a value for a single key on a frozen map"m
    auto key = _k->bind_and_get(params._options);
    auto value = _t->bind_and_get(params._options);
    if (!key) {
        throw invalid_request_exception("Invalid null map key");
    }
    auto ctype = static_cast<const map_type_impl*>(column.type.get());
    auto avalue = value ? params.make_cell(*ctype->get_values_type(), *value, atomic_cell::collection_member::yes) : params.make_dead_cell();
```

again we don't differentiate between unset/null

Reply:
Found another related bug.
When I fixed `maps::setter_by_key::execute`:
```
void
maps::setter_by_key::execute(mutation& m, const clustering_key_prefix& prefix, const update_parameters& params) {
    using exceptions::invalid_request_exception;
    assert(column.type->is_multi_cell()); // "Attempted to set a value for a single key on a frozen map"m
    auto key = _k->bind_and_get(params._options);
    auto value = _t->bind_and_get(params._options);
    if (!key) {
        throw invalid_request_exception("Invalid null map key");
    }
    if (value.is_unset_value()) {
        return;
    }
    auto ctype = static_cast<const map_type_impl*>(column.type.get());
    auto avalue = value ? params.make_cell(*ctype->get_values_type(), *value, atomic_cell::collection_member::yes) : params.make_dead_cell();
```
and did:
```
cqlsh:a> create table tmut (a int primary key, b map<int, frozen<ut>>);
cqlsh:a> update tmut set b[0] = {a:0, b:0} where a = 0;
cqlsh:a> select * from tmut;

 a | b
---+-------------------
 0 | {0: {a: 0, b: 0}}
```
then, from python driver:
```
    session.execute(session.prepare('update tmut set b[0] = ? where a = 0'))
```
guess what happened:
```
cqlsh:a> select * from tmut;

 a | b
---+---
```

But `map<int, int>` works correctly.

This is because `user_type::marker::bind_and_get` is inherited from `non_terminal`: 
```
        virtual cql3::raw_value_view bind_and_get(const query_options& options) override {
            auto t = bind(options);
            if (t) {
                return options.make_temporary(t->get(options));
            }
            return cql3::raw_value_view::make_null();
        };
```

`user_type::marker::bind` returns `constants::UNSET_VALUE`, but `options.make_temporary` changes this to null:
```
        cql3::raw_value_view query_options::make_temporary(cql3::raw_value value) const
        {
            if (value) {
                auto value_view = *value;
                auto ptr = _temporaries.write_place_holder(value_view.size());
                std::copy_n(value_view.data(), value_view.size(), ptr);
                return cql3::raw_value_view::make_value(fragmented_temporary_buffer::view(bytes_view{ptr, value_view.size()}));
            }
            return cql3::raw_value_view::make_null();
        }
```

changing `make_temporary` fixes the issue:
```
cql3::raw_value_view query_options::make_temporary(cql3::raw_value value) const
{
    if (value.is_null()) {
        return cql3::raw_value_view::make_null();
    } else if (value.is_unset_value()) {
        return raw_value_view::make_unset_value();
    }

    const auto& bytes = *value;
    auto ptr = _temporaries.write_place_holder(bytes.size());
    std::copy_n(bytes.data(), bytes.size(), ptr);
    return cql3::raw_value_view::make_value(fragmented_temporary_buffer::view(bytes_view{ptr, bytes.size()}));
}
```

Reply:
I switched to `bind_and_get` in `setter_by_field::execute`, but I didn't fix the bugs with other setters and make_temporary. Another PR is more suitable for that.

Reply:
btw, the whole unset thing is a huge mistake, it mixes semantics with the protocol.

It would have been much better to do it in the language itself:

    UPDATE tab
    SET col1 = ? CONDITIONAL ON ?, col2 = ? CONDIONAL ON ?
    WHERE pk = ? AND ck = ?

this allows the prepare phase to distinuish between a statement that can have unset values and a statement that cannot.